libr/core/cmd_search.c

```diff
@@ -1549,6 +1550,141 @@ static void do_esil_search(RCore *core, struct search_parameters *param, const c
 	r_cons_clear_line (1);
 }
 
+#define MAXINSTR 8
+#define SUMARRAY(arr, size, res) do (res) += (arr)[--(size)]; while ((size))
+
+static inline bool isnonlinear(int optype) {
+	return (optype ==  R_ANAL_OP_TYPE_CALL || optype ==  R_ANAL_OP_TYPE_JMP || optype ==  R_ANAL_OP_TYPE_CJMP);
+}	
+
+static void do_syscall_search(RCore *core, struct search_parameters *param) {
+	RSearch *search = core->search;
+	ut64 at, curpc;
+	ut8 *buf, *arr;
+	int previnstr[MAXINSTR + 1];
+	int curpos, idx, count = 0;
+	RAnalOp aop;
+	int i, j, inslen, ret, bsize = R_MIN (64, core->blocksize);
+	int kwidx = core->search->n_kws;
+	RAnalEsil *esil = core->anal->esil;
+	RIOMap* map;
+	RListIter *iter;
+	const int mininstrsz = r_anal_archinfo (core->anal, R_ANAL_ARCHINFO_MIN_OP_SIZE);
+        const int minopcode = R_MAX (1, mininstrsz);
+	int stacksize = r_config_get_i (core->config, "esil.stack.depth");
+	int iotrap = r_config_get_i (core->config, "esil.iotrap");
+	const char *a0 = r_reg_get_name (core->anal->reg, R_REG_NAME_SN);
+	const char *pc = r_reg_get_name (core->dbg->reg, R_REG_NAME_PC);
+	struct r_reg_item_t *r = r_reg_get (core->dbg->reg, pc, -1);
+	struct r_reg_item_t *reg_a0 = r_reg_get (core->dbg->reg, a0, -1);
+	
+	if (!(esil = r_anal_esil_new (stacksize, iotrap))) {
+		return;
+	}
+	memset (previnstr, 0, sizeof (previnstr));
+	buf = malloc (bsize);
+	arr = malloc (bsize);
+	if (!buf || !arr) {
+		eprintf ("Cannot allocate %d bytes\n", bsize);
+		free(buf);
+		free(arr);
+		return;
+	}
+	r_cons_break_push (NULL, NULL);
+	r_list_foreach (param->boundaries, iter, map) {
+		ut64 from = map->itv.addr;
+		ut64 to = r_itv_end (map->itv);
+		for (i = 0, at = from; at < to; at++, i++) {
+			if (r_cons_is_breaked ()) {
+				break;
+			}
+			if (i >= (bsize - 32)) {
+				i = 0;
+			}
+			if (!i) {
+				r_core_read_at (core, at, buf, bsize);
+			}
+			ret = r_anal_op (core->anal, &aop, at, buf + i, bsize - i);
+			curpos = idx++ % (MAXINSTR + 1);
+			previnstr[curpos] = ret; // This array holds prev n instr size + cur instr size
+			if (ret) {
+				bool match = false;
+				if (aop.type == R_ANAL_OP_TYPE_SWI) {
+					match = true;
+				}
+				if (match) {
```

Comment:
i would rather put that in a separate function and just do the following: if (aop.type == R_ANAL_OP_TYPE_SWI) { emulateSyscallPrelude(); 
}

passing the info needed to do that emulation, like RCore* , address to start emulating, etc. and return the syscall number